File: sdks/go/reproit/observation_session.go
package reproit

import "sync"

type observationAction string
type observationOutcome string
type observationSessionState uint8

const (
	observationCapture observationAction = "capture"
	observationReplay  observationAction = "replay"

	observationError    observationOutcome = "error"
	observationResponse observationOutcome = "response"

	observationRequestState observationSessionState = iota
	observationCaptureState
	observationReplayState
	observationReplayEOFState
	observationFinishedState
)

type installedObservationAdapter struct {
	adapterID            string
	adapterVersion       string
	class                automaticObservationClass
	implementationDigest string
}

type observationAdapterRegistry struct {
	mu       sync.RWMutex
	adapters map[automaticObservationClass]installedObservationAdapter
}

var installedObservationAdapters observationAdapterRegistry

func (registry *observationAdapterRegistry) install(adapter installedObservationAdapter) error {
	registry.mu.Lock()
	defer registry.mu.Unlock()
	if adapter.adapterID == "" || adapter.adapterVersion == "" || adapter.implementationDigest == "" ||
		len(registry.adapters) >= sdkEngineMaxObservationAdapters {
		return ErrAutomaticCapture
	}
	if registry.adapters == nil {
		registry.adapters = make(map[automaticObservationClass]installedObservationAdapter)
	}
	if _, exists := registry.adapters[adapter.class]; exists {
		return ErrAutomaticCapture
	}
	registry.adapters[adapter.class] = adapter
	return nil
}

func (registry *observationAdapterRegistry) remove(adapter installedObservationAdapter) {
	registry.mu.Lock()
	defer registry.mu.Unlock()
	if installed, exists := registry.adapters[adapter.class]; exists && installed == adapter {
		delete(registry.adapters, adapter.class)
	}
}

func (registry *observationAdapterRegistry) snapshot() []sdkEngineObservationAdapter {
	registry.mu.RLock()
	defer registry.mu.RUnlock()
	result := make([]sdkEngineObservationAdapter, 0, len(registry.adapters))
	for _, class := range sdkEngineRequiredObservationClasses {
		adapter, exists := registry.adapters[class]
		if !exists {
			continue
		}
		result = append(result, sdkEngineObservationAdapter{
			AdapterID: adapter.adapterID, AdapterVersion: adapter.adapterVersion,
			Class: string(adapter.class), ImplementationDigest: adapter.implementationDigest,
		})
	}
	return result
}

type observationSession struct {
	bridge          *sdkEngineBridge
	handle          sdkEngineObservationHandle
	mu              sync.Mutex
	requestBytes    uint64
	responseBytes   uint64
	sessionPosition uint64
	state           observationSessionState
}

func (operation *AutomaticOperation) openObservationSession(
	class automaticObservationClass,
	causalParentID *string,
) (*observationSession, error) {
	operation.mu.Lock()
	defer operation.mu.Unlock()
	if operation.finished || operation.worldComplete {
		return nil, ErrAutomaticCapture
	}
	started, err := operation.project.bridge.openObservation(
		operation.handle, string(class), causalParentID,
	)
	if err != nil {
		operation.abandonLocked()
		return nil, ErrAutomaticCapture
	}
	return &observationSession{
		bridge: operation.project.bridge, handle: started.Handle,
		sessionPosition: started.SessionPosition, state: observationRequestState,
	}, nil
}

func (session *observationSession) writeRequest(chunk []byte) error {
	return session.write("request", chunk, observationRequestState)
}

func (session *observationSession) writeResponse(chunk []byte) error {
	return session.write("response", chunk, observationCaptureState)
}

func (session *observationSession) write(
	stream string, chunk []byte, required observationSessionState,
) error {
	session.mu.Lock()
	defer session.mu.Unlock()
	if session.state != required || len(chunk) == 0 || len(chunk) > sdkEngineMaxObservationChunkBytes {
		return ErrAutomaticCapture
	}
	if session.bridge.writeObservation(session.handle, stream, chunk) != nil {
		session.abandonLocked()
		return ErrAutomaticCapture
	}
	if stream == "request" {
		session.requestBytes += uint64(len(chunk))
	} else {
		session.responseBytes += uint64(len(chunk))
	}
	return nil
}

func (session *observationSession) dispatch() (observationAction, error) {
	session.mu.Lock()
	defer session.mu.Unlock()
	if session.state != observationRequestState || session.requestBytes == 0 {
		return "", ErrAutomaticCapture
	}
	action, err := session.bridge.dispatchObservation(session.handle)
	if err != nil {
		session.abandonLocked()
		return "", ErrAutomaticCapture
	}
	switch observationAction(action) {
	case observationCapture:
		session.state = observationCaptureState
	case observationReplay:
		session.state = observationReplayState
	default:
		session.abandonLocked()
		return "", ErrAutomaticCapture
	}
	return observationAction(action), nil
}

func (session *observationSession) readResponse() ([]byte, bool, error) {
	session.mu.Lock()
	defer session.mu.Unlock()
	if session.state != observationReplayState {
		return nil, false, ErrAutomaticCapture
	}
	result, err := session.bridge.readObservation(session.handle)
	if err != nil {
		session.abandonLocked()
		return nil, false, ErrAutomaticCapture
	}
	if result.EOF {
		session.state = observationReplayEOFState
	}
	return result.Chunk, result.EOF, nil
}

func (session *observationSession) finish(outcome observationOutcome) error {
	session.mu.Lock()
	defer session.mu.Unlock()
	valid := session.state == observationCaptureState && session.responseBytes > 0
	valid = valid || session.state == observationReplayEOFState
	if !valid || (outcome != observationError && outcome != observationResponse) {
		return ErrAutomaticCapture
	}
	if session.bridge.finishObservation(
		session.handle, string(outcome), session.sessionPosition,
	) != nil {
		session.abandonLocked()
		return ErrAutomaticCapture
	}
	session.state = observationFinishedState
	return nil
}

func (session *observationSession) abandon() {
	session.mu.Lock()
	defer session.mu.Unlock()
	session.abandonLocked()
}

func (session *observationSession) abandonLocked() {
	if session.state == observationFinishedState {
		return
	}
	_ = session.bridge.abandonObservation(session.handle)
	session.state = observationFinishedState
}

File: sdks/dotnet/ReproIt.Sdk/ObservationSession.cs
using System.Text.Json.Nodes;

namespace ReproIt.Sdk;

internal enum ObservationAction
{
    Capture,
    Replay,
}

internal enum ObservationOutcome
{
    Error,
    Response,
}

internal enum ObservationSessionState
{
    Request,
    Capture,
    Replay,
    ReplayEof,
    Finished,
}

internal sealed record InstalledObservationAdapter(
    string AdapterId,
    string AdapterVersion,
    AutomaticObservationClass Class,
    string ImplementationDigest);

internal static class InstalledObservationAdapters
{
    private static readonly object RegistryLock = new();
    private static readonly SortedDictionary<AutomaticObservationClass, InstalledObservationAdapter>
        Registry = [];

    internal static void Install(InstalledObservationAdapter adapter)
    {
        lock (RegistryLock)
        {
            if (string.IsNullOrEmpty(adapter.AdapterId) ||
                string.IsNullOrEmpty(adapter.AdapterVersion) ||
                string.IsNullOrEmpty(adapter.ImplementationDigest) ||
                Registry.Count >= SdkEngineBridge.MaxObservationAdapters ||
                Registry.ContainsKey(adapter.Class))
            {
                throw AutomaticProject.CaptureError();
            }
            Registry.Add(adapter.Class, adapter);
        }
    }

    internal static void Remove(InstalledObservationAdapter adapter)
    {
        lock (RegistryLock)
        {
            if (Registry.TryGetValue(adapter.Class, out InstalledObservationAdapter? installed) &&
                installed == adapter)
            {
                Registry.Remove(adapter.Class);
            }
        }
    }

    internal static JsonArray Snapshot()
    {
        lock (RegistryLock)
        {
            JsonArray result = [];
            foreach (AutomaticObservationClass observationClass in
                SdkEngineBridge.RequiredObservationClasses)
            {
                if (!Registry.TryGetValue(observationClass, out InstalledObservationAdapter? adapter))
                {
                    continue;
                }
                result.Add(new JsonObject
                {
                    ["adapter_id"] = adapter.AdapterId,
                    ["adapter_version"] = adapter.AdapterVersion,
                    ["class"] = AutomaticOperation.ObservationClass(adapter.Class),
                    ["implementation_digest"] = adapter.ImplementationDigest,
                });
            }
            return result;
        }
    }
}

internal sealed class ObservationSession
{
    private readonly SdkEngineBridge bridge;
    private readonly SdkEngineObservationHandle handle;
    private readonly object stateLock = new();
    private ulong requestBytes;
    private ulong responseBytes;
    private readonly ulong sessionPosition;
    private ObservationSessionState state = ObservationSessionState.Request;

    internal ObservationSession(SdkEngineBridge bridge, SdkEngineObservationStart start)
    {
        this.bridge = bridge;
        handle = start.Handle;
        sessionPosition = start.SessionPosition;
    }

    internal void WriteRequest(byte[] chunk) =>
        Write("request", chunk, ObservationSessionState.Request);

    internal void WriteResponse(byte[] chunk) =>
        Write("response", chunk, ObservationSessionState.Capture);

    internal ObservationAction Dispatch()
    {
        lock (stateLock)
        {
            if (state != ObservationSessionState.Request || requestBytes == 0)
            {
                throw AutomaticProject.CaptureError();
            }
            try
            {
                string action = bridge.DispatchObservation(handle);
                state = action == "capture"
                    ? ObservationSessionState.Capture
                    : ObservationSessionState.Replay;
                return action == "capture" ? ObservationAction.Capture : ObservationAction.Replay;
            }
            catch (Exception)
            {
                AbandonLocked();
                throw AutomaticProject.CaptureError();
            }
        }
    }

    internal SdkEngineObservationChunk ReadResponse()
    {
        lock (stateLock)
        {
            if (state != ObservationSessionState.Replay)
            {
                throw AutomaticProject.CaptureError();
            }
            try
            {
                SdkEngineObservationChunk chunk = bridge.ReadObservation(handle);
                if (chunk.Eof)
                {
                    state = ObservationSessionState.ReplayEof;
                }
                return chunk;
            }
            catch (Exception)
            {
                AbandonLocked();
                throw AutomaticProject.CaptureError();
            }
        }
    }

    internal void Finish(ObservationOutcome outcome)
    {
        lock (stateLock)
        {
            bool valid = state == ObservationSessionState.Capture && responseBytes > 0;
            valid = valid || state == ObservationSessionState.ReplayEof;
            if (!valid)
            {
                throw AutomaticProject.CaptureError();
            }
            try
            {
                bridge.FinishObservation(
                    handle,
                    outcome == ObservationOutcome.Response ? "response" : "error",
                    sessionPosition);
                state = ObservationSessionState.Finished;
            }
            catch (Exception)
            {
                AbandonLocked();
                throw AutomaticProject.CaptureError();
            }
        }
    }

    internal void Abandon()
    {
        lock (stateLock)
        {
            AbandonLocked();
        }
    }

    private void Write(string stream, byte[] chunk, ObservationSessionState required)
    {
        lock (stateLock)
        {
            if (state != required || chunk.Length is 0 or > SdkEngineBridge.MaxObservationChunkBytes)
            {
                throw AutomaticProject.CaptureError();
            }
            try
            {
                bridge.WriteObservation(handle, stream, chunk);
                if (stream == "request")
                {
                    requestBytes += checked((ulong)chunk.Length);
                }
                else
                {
                    responseBytes += checked((ulong)chunk.Length);
                }
            }
            catch (Exception)
            {
                AbandonLocked();
                throw AutomaticProject.CaptureError();
            }
        }
    }

    private void AbandonLocked()
    {
        if (state == ObservationSessionState.Finished)
        {
            return;
        }
        try
        {
            bridge.AbandonObservation(handle);
        }
        catch (Exception)
        {
            // Capture cleanup must not change application behavior.
        }
        state = ObservationSessionState.Finished;
    }
}
